File: src/bc/BC.cpp
/*
  Copyright (C) 2012,2013
      Max Planck Institute for Polymer Research
  Copyright (C) 2008,2009,2010,2011
      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI
  
  This file is part of ESPResSo++.
  
  ESPResSo++ is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
  
  ESPResSo++ is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>. 
*/

#include <python.hpp>
#include <boost/python/tuple.hpp>

#include "BC.hpp"
#include "Real3D.hpp"
#include "Int3D.hpp"
#include "esutil/RNG.hpp"

namespace espressopp {
  namespace bc {

    LOG4ESPP_LOGGER(BC::logger, "BC");

    Real3D
    BC::getMinimumImageVector(const Real3D& pos1,
			      const Real3D& pos2) const {
      Real3D res;
      getMinimumImageVector(res, pos1, pos2);
      return res;
    }

    Real3D
    BC::getRandomPos() const {
      Real3D res;
      getRandomPos(res);
      return res;
    }

    void 
    BC::foldPosition(Real3D& pos, Int3D& imageBox) const {
      for (int i = 0; i < 3; ++i)
	foldCoordinate(pos, imageBox, i);
    }

    void
    BC::foldPosition(Real3D& pos) const {
      Int3D imageBox(0, 0, 0);
      foldPosition(pos, imageBox);
    }

    void 
    BC::unfoldPosition(Real3D& pos, Int3D& imageBox) const {
      for (int i = 0; i < 3; ++i)
	unfoldCoordinate(pos, imageBox, i);
    }

    boost::python::tuple 
    BC::getFoldedPosition(const Real3D& pos, 
			  const Int3D& imageBox) const {
      Real3D foldedPos = pos;
      Int3D foldedImageBox = imageBox;
      foldPosition(foldedPos, foldedImageBox);
      return boost::python::make_tuple(foldedPos, foldedImageBox);
    }

    boost::python::tuple 
    BC::getFoldedPosition(const Real3D& pos) const {
      Real3D foldedPos = pos;
      Int3D foldedImageBox(0, 0, 0);
      foldPosition(foldedPos, foldedImageBox);
      return boost::python::make_tuple(foldedPos, foldedImageBox);
    }

    Real3D
    BC::getUnfoldedPosition(const Real3D& pos, 
			    const Int3D& imageBox) const {
      Real3D unfoldedPos = pos;
      Int3D unfoldedImageBox = imageBox;
      unfoldPosition(unfoldedPos, unfoldedImageBox);
      return unfoldedPos;
    }

    //////////////////////////////////////////////////
    // REGISTRATION WITH PYTHON
    //////////////////////////////////////////////////
    void
    BC::registerPython() {
      using namespace espressopp::python;
    
      // also register the abstract class BC to make virtual functions available
      // be careful: boost::noncopyable must be used for abstract classes with pure routines
      // no_init must be used as the abstract class BC has no constructor

      Real3D (BC::*pygetMinimumImageVector)(const Real3D& pos1,
					    const Real3D& pos2) const 
	= &BC::getMinimumImageVector;
      
      boost::python::tuple (BC::*pygetFoldedPosition1)
	(const Real3D& pos, const Int3D& imageBox) const
	= &BC::getFoldedPosition;

      boost::python::tuple (BC::*pygetFoldedPosition2)
	(const Real3D& pos) const
	= &BC::getFoldedPosition;

      Real3D (BC::*pygetUnfoldedPosition)
	(const Real3D& pos, const Int3D& imageBox) const 
	= &BC::getUnfoldedPosition;

      Real3D (BC::*pygetRandomPos)() const = &BC::getRandomPos;
    
      class_< BC, boost::noncopyable >("bc_BC", no_init)
	.add_property("boxL", &BC::getBoxL)
	.add_property("rng", &BC::getRng, &BC::setRng)
	.def("getMinimumImageVector", pygetMinimumImageVector)
	.def("getFoldedPosition", pygetFoldedPosition1)
	.def("getFoldedPosition", pygetFoldedPosition2)
	.def("getUnfoldedPosition", pygetUnfoldedPosition)
	.def("getRandomPos", pygetRandomPos)
	;
    }
  }
}

File: src/bc/BC.py
#  Copyright (C) 2012,2013
#      Max Planck Institute for Polymer Research
#  Copyright (C) 2008,2009,2010,2011
#      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI
#  
#  This file is part of ESPResSo++.
#  
#  ESPResSo++ is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  
#  ESPResSo++ is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>. 


r"""

This is the abstract base class for all boundary condition objects.
It cannot be used directly. All derived classes implement at least
the following methods:

.. py:class:: espressopp.bc.BC

	.. py:method:: espressopp.bc.BC.getFoldedPosition(pos, imageBox)

		:param pos: 
		:param imageBox: (default: None)
		:type pos: 
		:type imageBox: 
		:rtype: 

	.. py:method:: espressopp.bc.BC.getMinimumImageVector(pos1, pos2)

		:param pos1: 
		:param pos2: 
		:type pos1: 
		:type pos2: 
		:rtype: 

	.. py:method:: espressopp.bc.BC.getRandomPos()

		:rtype: 

	.. py:method:: espressopp.bc.BC.getUnfoldedPosition(pos, imageBox)

		:param pos: 
		:param imageBox: 
		:type pos: 
		:type imageBox: 
		:rtype: 

	`pos`, `pos1` and `pos2` are particle coordinates ( type: (`float`, `float`, `float`) ).
	`imageBox` ( type: (`int`, `int`, `int`) ) specifies the   

"""


from espressopp import pmi
from espressopp import toReal3D, toReal3DFromVector, toInt3D, toInt3DFromVector
from _espressopp import bc_BC 

class BCLocal(object):
    def getMinimumImageVector(self, pos1, pos2):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup() or pmi.isController:
            return self.cxxclass.getMinimumImageVector(
                self, toReal3DFromVector(pos1), toReal3DFromVector(pos2))

    def getFoldedPosition(self, pos, imageBox=None):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup() or pmi.isController:
            if imageBox is None:
                return self.cxxclass.getFoldedPosition(self, toReal3DFromVector(pos))
            else:
                return self.cxxclass.getFoldedPosition(
                    self, toReal3DFromVector(pos), toInt3DFromVector(imageBox))

    def getUnfoldedPosition(self, pos, imageBox):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup() or pmi.isController:
            return self.cxxclass.getUnfoldedPosition(
                self, toReal3DFromVector(pos), toInt3DFromVector(imageBox))

    def getRandomPos(self):
        if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup() or pmi.isController:
            return self.cxxclass.getRandomPos(self)
    
if pmi.isController :
    class BC(object):
        __metaclass__ = pmi.Proxy
        pmiproxydefs = dict(
            pmiproperty = [ "boxL", "rng" ],
            localcall = [ "getMinimumImageVector", 
                          "getFoldedPosition", "getUnfoldedPosition", 
                          "getRandomPos" ]
            )
